### PR TITLE
CI: Support up and downstream dependencies

### PR DESCRIPTION
**Pull request by ColdenCullen**
With this PR, each repo may define it's own upstream dependencies, and downstream projects to test (when running a downstream build).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .codebuild.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (justinboswell) [inline .codebuild.json]: My only concern with this file is the name, who's to say codebuild doesn't eventually have a .codebuild.json default file. Maybe just name it something different?

File: codebuild/builder.py [hunk 2]
```diff
@@ -490,24 +497,150 @@ def run_build(build_spec, is_dryrun):
...
+        _log_command(*command)
         if not is_dryrun:
-            subprocess.check_call(_flatten_command(command), stdout=sys.stdout, stderr=sys.stderr)
+            subprocess.check_call(_flatten_command(*command), stdout=sys.stdout, stderr=sys.stderr)
+
+    # Helper to run makedirs regardless of dry run status
+    def _mkdir(directory):
+        if not is_dryrun:
+            os.makedirs(directory, exist_ok=True)
+        _log_command("mkdir", "-p", directory)
```
Comment (justinboswell) [inline codebuild/builder.py]: Can we log before attempting the makedirs, so if an exception is thrown, we have some info?

File: codebuild/builder.py [hunk 3]
```diff
@@ -490,24 +497,150 @@ def run_build(build_spec, is_dryrun):
...
+    def _cd(directory):
+        if is_dryrun:
+            global cwd
+            if os.path.isabs(directory) or directory.startswith('$'):
+                cwd = directory
+            else:
+                cwd = os.path.join(cwd, directory)
+        else:
+            os.chdir(directory)
+        _log_command("cd", directory)
```
Comment (justinboswell) [inline codebuild/builder.py]: Same here, log first.

File: codebuild/builder.py [hunk 4]
```diff
@@ -516,23 +649,27 @@ def _run_command(command):
         build_dir = tempfile.mkdtemp()
     _log_command(['mkdir', build_dir])
```
Comment (justinboswell) [inline codebuild/builder.py]: Log first

File: codebuild/builder.py (hunk 2 above)
Comment (ColdenCullen) [reply codebuild/builder.py]: Sure

File: codebuild/builder.py (hunk 3 above)
Comment (ColdenCullen) [reply codebuild/builder.py]: Yup!

File: codebuild/builder.py (hunk 4 above)
Comment (ColdenCullen) [reply codebuild/builder.py]: Can't here, don't have the directory name until mkdtemp() returns.

**Pull request merged**